components/mpas-ocean/src/shared/mpas_ocn_vmix.F

```diff
@@ -925,6 +917,194 @@ subroutine ocn_vel_vmix_tend_implicit_spatially_variable_mannings(forcingPool, b
+      !
+      !-----------------------------------------------------------------
+
+      real (kind=RKIND), dimension(:,:), intent(inout) :: &
+         normalVelocity             !< Input: velocity
+
+      real (kind=RKIND), dimension(:,:), intent(inout) :: &
+         vertViscTopOfEdge !< Input: vertical mixing coefficients
+
+      !-----------------------------------------------------------------
+      !
+      ! output variables
+      !
+      !-----------------------------------------------------------------
+
+      integer, intent(out) :: err !< Output: error flag
+
+      !-----------------------------------------------------------------
+      !
+      ! local variables
+      !
+      !-----------------------------------------------------------------
+
+      integer :: iEdge, k, cell1, cell2, N, Nsurf
+
+      real (kind=RKIND), dimension(:), allocatable :: bTemp, C, rTemp, CdTemp
+      real (kind=RKIND) :: A, m
+
+      real (kind=RKIND) :: von_karman, von_karman_sq
+      real (kind=RKIND) :: kineticEnergyTopOfEdge, zMidEdge
+      integer :: iCell
+
+      err = 0
+      von_karman = 0.4_RKIND
+      von_karman_sq = 0.16_RKIND
+
+      if(.not.velVmixOn) return
+
+      allocate(bTemp(nVertLevels),C(nVertLevels),rTemp(nVertLevels),CdTemp(nVertLevels))
+
+#ifdef MPAS_OPENACC
+      !$acc enter data create(bTemp, C, rTemp)
+
+      !$acc parallel loop present(maxLevelEdgeTop, minLevelEdgeBot, cellsOnEdge, &
+      !$acc    layerThickEdgeMean, vertViscTopOfEdge, normalVelocity, &
+      !$acc    kineticEnergyCell, bottomDepth) &
+      !$acc    private(Nsurf, N, cell1, cell2, CdTemp, A, bTemp, C, m, rTemp, k)
+#else
+      !$omp do schedule(runtime)
+#endif
+      do iEdge = 1, nEdgesOwned
+        Nsurf = minLevelEdgeBot(iEdge)
+        N = maxLevelEdgeTop(iEdge)
+        if (N .gt. 0) then
+
+         cell1 = cellsOnEdge(1,iEdge)
+         cell2 = cellsOnEdge(2,iEdge)
+
+         CdTemp = 0.0_RKIND
+         do k = N, Nsurf, -1
+           zMidEdge = 0.5_RKIND * (bottomDepth(cell1) + bottomDepth(cell2) + zMid(k,cell1) + zMid(k,cell2))
+           if (k < N .and. zMidEdge > zDrag_max) exit
+           CdTemp(k) = min(implicitCd_max, &
+                        max(implicitCd_min, &
+                            von_karman_sq / &
+                            (log(1.0_RKIND + &
+                                 (zMidEdge/bottom_roughness)))**2))
+           ! For some reason we can use k-1 even when N=Nsurf
+           kineticEnergyTopOfEdge = 0.5_RKIND * (kineticEnergyCell(k,cell1) + kineticEnergyCell(k,cell2) + &
+                                                 kineticEnergyCell(k-1,cell1) + kineticEnergyCell(k-1,cell2))
+           vertViscTopOfEdge(k,iEdge) = max(vertViscTopOfEdge(k,iEdge), &
+                                            sqrt(CdTemp(k) * kineticEnergyTopOfEdge) * &
+                                            von_karman * (zMidEdge + bottom_roughness) &
+                                           )
```

Comment:
@dengwirda I just wanted to note that I made a decision to take the maximum value of `vertViscTopOfEdge` that results from the vertical mixing scheme or the vertical viscosity that results from bottom drag. I think this will result in better bottom boundary layer velocity profiles (when resolved). I also saw that with many vertical layers the drag at the bottommost layer tends to be quite strong and I didn't want to enhance this even further by adding `vertViscTopOfEdge` to the drag solution, which would be the other sensible way to approach this.

Reply:
Thanks @cbegeman, this makes a lot of sense to me --- I think in cases with high enough vertical resolution the layers inside the bottom BL probably shouldn't get a contribution from the vmix scheme as well as from the log profile.